File: uva_10006.cpp
# include <iostream>
using namespace std;
long long int mod(long long int a, long long int n){
    long long int N=n, ans, temp;
    if(n%2)
        ans=a;
    else
        ans=1;
    while(n/2){
        //cout << n/2 << endl;
        n/=2;
        temp=(a*a)%N;
        if (n%2)
            ans=(ans*temp)%N;
        a=temp;

    }
    return ans;
}
bool check(long long int a, long long int n){
    if(mod(a, n)==a)
        return true;
    else
        return false;
}
int main(void){
    long long int n;
    bool prime[65001] = {true, true, false};
    for (int i = 2; i <= 65000; ++i)
        if (!prime[i])
            for (int j = i+i; j <= 65000; j+=i)
                prime[j] = true;
    while(cin >> n){
        if (!n)
            break;
        bool isCarmichael = true;
        if (!prime[n])
            isCarmichael = false;

        for(long long int i=2; i < n && isCarmichael; ++i)
            isCarmichael = isCarmichael && check(i, n);  

        if(isCarmichael)
            cout << "The number " << n << " is a Carmichael number.\n";
        else
            cout << n << " is normal.\n";
    }
    return 0;
}

File: uva_10006.py
def mod(a, n):
    N=n
    ans=a if n%2 else 1
    while int(n/2):
        n=int(n/2)
        temp=(a*a)%N
        if n%2:
            ans=(ans*temp)%N
        a=temp
    return ans

def check(a, n):
    if mod(a, n)==a:
        return True
    else:
        return False

prime=[False for _ in range(65001)]
prime[0], prime[1] = True, True
for i in range(2, 65001):
    if not prime[i]:
        j=2*i
        while j<=65000:
            prime[j]=True
            j+=i

while True:
    try:
        n=int(input())
    except:
        break
    if not n:
        break
    isCarmichael = True
    if not prime[n]:
        isCarmichael = False

    for i in range(2, n):
        isCarmichael = isCarmichael and check(i, n)
        if not isCarmichael:
            break

    if isCarmichael:
        print("The number {} is a Carmichael number.".format(n))
    else:
        print("{} is normal.".format(n))